### PR TITLE
Introduce <botan/boost_compat.h> as feature flag

### PR DESCRIPTION
**Pull request by reneme**
This introduces `#include <botan/boost_compat.h>` that defines `BOTAN_MINIMUM_SUPPORTED_BOOST_VERSION` as the minimum version of boost to be usable by this version of Botan. If the available boost headers are new enough, `BOTAN_FOUND_COMPATIBLE_BOOST_VERSION` is defined as well. In this case, it is safe to `#include <botan/asio_stream.h>`, otherwise a macro-error would be generated.

Usage example:

```C++
#include <botan/boost_compat.h>

#if defined(BOTAN_FOUND_COMPATIBLE_BOOST_VERSION)
#include <botan/asio_stream.h>
// profit...
#else
// fallback...
#endif

```

### CODE REVIEW

File: src/tests/unit_asio_stream.cpp [hunk 1]
```diff
@@ -736,5 +736,5 @@ BOTAN_REGISTER_TEST("tls", "tls_asio_stream", Asio_Stream_Tests);
 
 }  // namespace Botan_Tests
 
-   #endif  // BOOST_VERSION
-#endif     // BOTAN_HAS_TLS && BOTAN_HAS_BOOST_ASIO
+   #endif
+#endif  // BOTAN_HAS_TLS && BOTAN_HAS_BOOST_ASIO
```
Comment (lieser) [inline src/tests/unit_asio_stream.cpp:740]:
```suggestion
#endif  // BOTAN_HAS_TLS && BOTAN_HAS_TLS_ASIO_STREAM
```

File: src/lib/utils/boost/boost_compat.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    *
+    * until Botan 3.2.0
+    *   1.66.0 - first version to be compatible with Networking TS (N4656) and boost::beast
+    *
+    * as of Botan 3.3.0
+    *   1.73.0 - first version supporting the C++20 concepts syntax
+    */
+   #define BOTAN_MINIMUM_SUPPORTED_BOOST_VERSION 107300
+
+   #include <boost/version.hpp>
```
Comment (lieser) [inline src/lib/utils/boost/boost_compat.h:30]: Move include before the define?

File: src/lib/utils/boost/boost_compat.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    *  BOOST_VERSION / 100 % 1000 is the minor version
+    *  BOOST_VERSION / 100000 is the major version
+    *
+    * until Botan 3.2.0
+    *   1.66.0 - first version to be compatible with Networking TS (N4656) and boost::beast
+    *
+    * as of Botan 3.3.0
+    *   1.73.0 - first version supporting the C++20 concepts syntax
+    */
+   #define BOTAN_MINIMUM_SUPPORTED_BOOST_VERSION 107300
```
Comment (randombit) [inline src/lib/utils/boost/boost_compat.h:28]: This comment should be clarified that it only applies to the asio integrations; we can still use older Boost for eg socket wrappers.

**Review by randombit — APPROVED** [review]
Left a couple of minor comments (and I concur on the comments @lieser left) but basically LGTM.

Can you please amend the relevant docs/examples?

File: src/tests/unit_asio_stream.cpp (hunk 1 above)
Comment (reneme) [reply src/tests/unit_asio_stream.cpp:740]: Mhh, somehow, I thought `clang-format` was responsible for that. Fixed.

**Comment by reneme** [comment]
I applied the suggestions, extended `src/examples/tls_stream_client.cpp` and added a note about `BOTAN_FOUND_COMPATIBLE_BOOST_VERSION` in the documentation.

**Review by lieser — APPROVED** [review]

**Pull request merged**